src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,201 @@
+            0x00 /*body*/});
+
+    internal static string WebTransportProtocolValue => "webtransport";
+    internal static string SecPrefix => "sec-webtransport-http3-";
+    internal static string VersionHeaderPrefix => $"{SecPrefix}draft";
+    internal static string CurrentSuppportedVersion => $"{VersionHeaderPrefix}02";
+
+    long IWebTransportSession.SessionId => _connectStream.StreamId;
+
+    internal WebTransportSession(Http3Connection connection, Http3Stream connectStream)
+    {
+        _connection = connection;
+        _connectStream = connectStream;
+        _isClosing = false;
+        // unbounded as limits to number of streams is enforced elsewhere
+        _pendingStreams = Channel.CreateUnbounded<WebTransportStream>();
+
+        // listener to abort if this connection is closed
+        _connectionClosedRegistration = connection._multiplexedContext.ConnectionClosed.Register(state =>
+        {
+            var session = (WebTransportSession)state!;
+            session.OnClientConnectionClosed();
+        }, this);
+    }
+
+    void IWebTransportSession.Abort(int errorCode)
+    {
+        Abort(new(), (Http3ErrorCode)errorCode);
+    }
+
+    internal void OnClientConnectionClosed()
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+
+        _connectionClosedRegistration.Dispose();
+
+        lock (_openStreams)
+        {
+            foreach (var stream in _openStreams)
+            {
+                _ = stream.Value.DisposeAsync().AsTask();
+            }
+
+        }
+        _openStreams.Clear();
+
+        _pendingStreams.Writer.Complete();
+    }
+
+    internal void Abort(ConnectionAbortedException exception, Http3ErrorCode error)
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+
+        _connectionClosedRegistration.Dispose();
+
+        lock (_openStreams)
+        {
+            _connectStream.Abort(exception, error);
+            foreach (var stream in _openStreams)
+            {
+                if (exception.InnerException is not null)
+                {
+                    stream.Value.Abort(new ConnectionAbortedException(exception.Message, exception.InnerException));
+                }
+                else
+                {
+                    stream.Value.Abort(new ConnectionAbortedException(exception.Message));
+                }
+            }
+        }
+
+        _openStreams.Clear();
+
+        _pendingStreams.Writer.Complete();
+    }
+
+    public async ValueTask<ConnectionContext> OpenUnidirectionalStreamAsync(CancellationToken cancellationToken)
+    {
+        if (_isClosing)
+        {
+            throw new ObjectDisposedException(CoreStrings.WebTransportIsClosing);
+        }
+        // create the stream
+        var features = new FeatureCollection();
+        features.Set<IStreamDirectionFeature>(new DefaultStreamDirectionFeature(canRead: false, canWrite: true));
```

Comment:
The `FeatureCollection` and `DefaultStreamDirectionFeature` seems like it can be a static since it's always the same. (unless of course the features are modified elsewhere, in which case at least the `DefaultStreamDirectionFeature` can be static)

Reply:
I think that this might cause a ton of confusion. Is there a way to toggle the `IsReadOnly` property? I see that currently it is set to be false at all times. I am just worried that someone modifies it and then breaks something indirectly. We are encouraging the application to use these feature collections. So, I think their robustness is very important.